src/ucs/sys/topo/base/topo.c

```diff
@@ -790,3 +789,52 @@ const char *ucs_topo_resolve_sysfs_path(const char *dev_path, char *path_buffer)
               sysfs_path);
     return sysfs_path;
 }
+
+static ucs_status_t
+uct_topo_iface_parse_virtual_dev(const struct dirent *entry, void *ctx)
+{
+    static const char *low_level_dev_prefix = "lower_";
+    ucs_string_buffer_t *dev_path           = ctx;
+
+    if (!strncmp(entry->d_name, low_level_dev_prefix,
+                 strlen(low_level_dev_prefix))) {
+        ucs_string_buffer_appendf(dev_path, "/%s", entry->d_name);
+        return UCS_ERR_CANCELED;
+    }
+
+    return UCS_OK;
+}
+
+const char *ucs_topo_iface_lowest_device_path(const char *dev_name,
+                                              const char *dev_type_root,
+                                              char *lowest_path_buffer,
+                                              int buffer_length)
+{
+    ucs_string_buffer_t dev_path = UCS_STRING_BUFFER_INITIALIZER;
+    const int max_depth          = 8;
+    int depth;
+    const char *lowest_path;
+    ucs_status_t status;
+    
+    ucs_string_buffer_appendf(&dev_path, "%s/%s", dev_type_root, dev_name);
+
+    for (depth = 0; depth < max_depth; ++ depth) {
+        status = ucs_sys_readdir(ucs_string_buffer_cstr(&dev_path),
+                                 uct_topo_iface_parse_virtual_dev, &dev_path);
+        if (status != UCS_ERR_CANCELED) {
+            break;
+        }
+    }
+
+    if (status != UCS_OK) {
+        ucs_string_buffer_cleanup(&dev_path);
+        return NULL;
+    }
+
+    strncpy(lowest_path_buffer, ucs_string_buffer_cstr(&dev_path), buffer_length - 1);
+    lowest_path_buffer[buffer_length - 1] = '\0';
```

Comment:
replace with `ucs_strncpy_safe` and remove next line (`lowest_path_buffer[buffer_length - 1] = '\0';`)